### PR TITLE
Throw when setting Frame.StatusCode after response has already started (#805).

### PR DESCRIPTION
**Pull request by cesarbs**
Note this doesn't change the response. Since the response has already started we can't send a 500. At least we get a log message.

#805 

cc @halter73 @mikeharder @Tratcher @pakrym

### CODE REVIEW

File: src/Microsoft.AspNetCore.Server.Kestrel/Http/Frame.cs [hunk 1]
```diff
@@ -115,7 +115,24 @@ public string HttpVersion
...
+                if (HasResponseStarted)
+                {
+                    throw new InvalidOperationException("Status code cannot be set, response has already started.");
+                }
+
+                _statusCode = value;
+            }
+        }
+
         public string ReasonPhrase { get; set; }
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Server.Kestrel/Http/Frame.cs]: ReasonPhrase too.

File: src/Microsoft.AspNetCore.Server.Kestrel/Http/Frame.cs [hunk 2]
```diff
@@ -115,7 +115,24 @@ public string HttpVersion
         public IHeaderDictionary RequestHeaders { get; set; }
         public Stream RequestBody { get; set; }
 
-        public int StatusCode { get; set; }
+        private int _statusCode;
+        public int StatusCode
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Server.Kestrel/Http/Frame.cs]: What about https://github.com/aspnet/KestrelHttpServer/blob/c6865f0e447b99bb5b56a943d155d986a4120bcb/src/Microsoft.AspNetCore.Server.Kestrel/Http/Frame.cs#L614-L617 ?

File: test/Microsoft.AspNetCore.Server.KestrelTests/FrameTests.cs [hunk 3]
```diff
@@ -50,5 +52,46 @@ public void EmptyHeaderValuesCanBeParsed(string rawHeaders, int numHeaders)
...
+
+            // Act
+            frame.Reset();
+
+            // Assert
+            Assert.Equal("http", ((IFeatureCollection)frame).Get<IHttpRequestFeature>().Scheme);
+        }
+
+        [Fact]
+        public void ThrowsWhenStatusCodeIsSetAfterResponseStarted()
```
Comment (Tratcher) [inline test/Microsoft.AspNetCore.Server.KestrelTests/FrameTests.cs]: Are there similar tests for setting Header values?

Comment (cesarbs) [inline test/Microsoft.AspNetCore.Server.KestrelTests/FrameTests.cs]: I don't think so. Will add.

File: src/Microsoft.AspNetCore.Server.Kestrel/Http/Frame.cs (hunk 2 above)
Comment (halter73) [inline src/Microsoft.AspNetCore.Server.Kestrel/Http/Frame.cs]: I would probably just change that to something like:


    if (HasResponseStarted)
    {
        // We can no longer respond with a 500, so we simply close the connection.
        _requestProcessingStopping = true;
        return TaskUtilities.CompletedTask;
    }
    else
    {
        if (_requestRejected)
        {
            // 400 Bad Request
            StatusCode = 400;
            _keepAlive = false;
        }
        else
        {
            // 500 Internal Server Error
            StatusCode = 500;
        }

        // ...

File: test/Microsoft.AspNetCore.Server.KestrelTests/FrameTests.cs (hunk 3 above)
Comment (halter73) [inline test/Microsoft.AspNetCore.Server.KestrelTests/FrameTests.cs]: Add it to `FrameResponseHeadersTests`.

File: src/Microsoft.AspNetCore.Server.Kestrel/Http/Frame.cs (hunk 2 above)
Comment (cesarbs) [inline src/Microsoft.AspNetCore.Server.Kestrel/Http/Frame.cs]: Yeah, it's what I did.

File: test/Microsoft.AspNetCore.Server.KestrelTests/FrameTests.cs (hunk 3 above)
Comment (cesarbs) [inline test/Microsoft.AspNetCore.Server.KestrelTests/FrameTests.cs]: Done.

**Comment by cesarbs** [comment]
Ping.

**Comment by halter73** [comment]
:shipit:

**Pull request merged**